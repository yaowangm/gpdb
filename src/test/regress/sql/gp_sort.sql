-- GPDB used to define Datum as a signed type. That caused below
-- soring issues.
create table mac_tbl (id int, mac macaddr) distributed by (id);
insert into mac_tbl values
(0, '00:00:00:00:00:00'),
(0, 'ff:ff:ff:ff:ff:ff'),
(0, '11:11:11:11:11:11');
select * from mac_tbl order by mac;
create table uuid_tbl (id int, uid uuid) distributed by (id);
insert into uuid_tbl values
(0, '00000000000000000000000000000000'),
(0, 'ffffffffffffffffffffffffffffffff'),
(0, '11111111111111111111111111111111');
select * from uuid_tbl order by uid;

-- Test cases for GP's specical mksort
-- start_ignore
create language plpythonu;
-- end_ignore

-- compare the results between mksort and qsort
create or replace function gp_mksort_test() returns int as $$
# the compare utility function
def compare_results(query):
    plpy.execute("set gp_enable_mk_sort to on;")
    mksort_res = plpy.execute(query)
    plpy.execute("set gp_enable_mk_sort to off;")
    #plpy.info(plpy.execute("show gp_enable_mk_sort;")[0]["gp_enable_mk_sort"])
    qsort_res = plpy.execute(query)
    res1 = mksort_res
    res2 = qsort_res
    if res1.nrows() != res2.nrows():
        plpy.error("gp_mksort_test: query count failed")

    for i in range(res1.nrows()):
        if  res1[i]['a'] != res2[i]['a'] or \
            res1[i]['b'] != res2[i]['b'] or \
            res1[i]['c'] != res2[i]['c']:
            plpy.error("gp_mksort_test: query results failed")

# basic test: 1~13 rows
for i in range(13):
    plpy.execute("truncate gp_mksort_test_table;")
    insert = "insert into gp_mksort_test_table \
        select floor(random()*10), floor(random()*100), left(md5(g::text),4) \
        from generate_series(1,%d) g;" % (i+1)
    plpy.execute(insert)
    compare_results("select * from gp_mksort_test_table order by a,c")
plpy.info('gp_mksort_test: basic test passed')

# random test: run 10 times (10w rows)
for i in range(10):
    plpy.execute("truncate gp_mksort_test_table;")
    insert = "insert into gp_mksort_test_table \
        select floor(random()*10), floor(random()*100), left(md5(g::text),4) \
        from generate_series(1,100000) g;"
    plpy.execute(insert)
    compare_results("select * from gp_mksort_test_table order by a,b,c")
plpy.info('gp_mksort_test: random test passed')

return 0
$$ language plpythonu;

create table gp_mksort_test_table(a int, b int, c varchar);
set statement_mem='100MB';
select gp_mksort_test();

drop table gp_mksort_test_table;
drop function gp_mksort_test();
