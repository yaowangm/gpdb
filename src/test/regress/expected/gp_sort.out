-- GPDB used to define Datum as a signed type. That caused below
-- soring issues.
create table mac_tbl (id int, mac macaddr) distributed by (id);
insert into mac_tbl values
(0, '00:00:00:00:00:00'),
(0, 'ff:ff:ff:ff:ff:ff'),
(0, '11:11:11:11:11:11');
select * from mac_tbl order by mac;
 id |        mac        
----+-------------------
  0 | 00:00:00:00:00:00
  0 | 11:11:11:11:11:11
  0 | ff:ff:ff:ff:ff:ff
(3 rows)

create table uuid_tbl (id int, uid uuid) distributed by (id);
insert into uuid_tbl values
(0, '00000000000000000000000000000000'),
(0, 'ffffffffffffffffffffffffffffffff'),
(0, '11111111111111111111111111111111');
select * from uuid_tbl order by uid;
 id |                 uid                  
----+--------------------------------------
  0 | 00000000-0000-0000-0000-000000000000
  0 | 11111111-1111-1111-1111-111111111111
  0 | ffffffff-ffff-ffff-ffff-ffffffffffff
(3 rows)

-- Test cases for GP's specical mksort
-- start_ignore
create language plpython3u;
-- end_ignore
-- compare the results between mksort and qsort
create or replace function gp_mksort_test() returns int as $$
# the compare utility function
def compare_results(query):
    plpy.execute("set gp_enable_mk_sort to on;")
    mksort_res = plpy.execute(query)
    plpy.execute("set gp_enable_mk_sort to off;")
    #plpy.info(plpy.execute("show gp_enable_mk_sort;")[0]["gp_enable_mk_sort"])
    qsort_res = plpy.execute(query)
    res1 = mksort_res
    res2 = qsort_res
    if res1.nrows() != res2.nrows():
        plpy.error("gp_mksort_test: query count failed")

    for i in range(res1.nrows()):
        if  res1[i]['a'] != res2[i]['a'] or \
            res1[i]['b'] != res2[i]['b'] or \
            res1[i]['c'] != res2[i]['c']:
            plpy.error("gp_mksort_test: query results failed")

# basic test: 1~13 rows
for i in range(13):
    plpy.execute("truncate gp_mksort_test_table;")
    insert = "insert into gp_mksort_test_table \
        select floor(random()*10), floor(random()*100), left(md5(g::text),4) \
        from generate_series(1,%d) g;" % (i+1)
    plpy.execute(insert)
    compare_results("select * from gp_mksort_test_table order by a,c")
plpy.info('gp_mksort_test: basic test passed')

# random test: run 10 times (10w rows)
for i in range(10):
    plpy.execute("truncate gp_mksort_test_table;")
    insert = "insert into gp_mksort_test_table \
        select floor(random()*10), floor(random()*100), left(md5(g::text),4) \
        from generate_series(1,100000) g;"
    plpy.execute(insert)
    compare_results("select * from gp_mksort_test_table order by a,b,c")
plpy.info('gp_mksort_test: random test passed')

return 0
$$ language plpythonu;
create table gp_mksort_test_table(a int, b int, c varchar);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
set statement_mem='100MB';
select gp_mksort_test();
INFO:  gp_mksort_test: basic test passed
INFO:  gp_mksort_test: random test passed
 gp_mksort_test 
----------------
              0
(1 row)

drop table gp_mksort_test_table;
drop function gp_mksort_test();
